src/compiler/program.ts

```diff
@@ -1996,28 +2046,63 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
 
     onProgramCreateComplete();
 
-    // Add file processingDiagnostics
-    fileProcessingDiagnostics?.forEach(diagnostic => {
-        switch (diagnostic.kind) {
-            case FilePreprocessingDiagnosticsKind.FilePreprocessingFileExplainingDiagnostic:
-                return programDiagnostics.add(createDiagnosticExplainingFile(diagnostic.file && getSourceFileByPath(diagnostic.file), diagnostic.fileProcessingReason, diagnostic.diagnostic, diagnostic.args || emptyArray));
-            case FilePreprocessingDiagnosticsKind.FilePreprocessingReferencedDiagnostic:
-                const { file, pos, end } = getReferencedFileLocation(program, diagnostic.reason) as ReferenceFileLocation;
-                return programDiagnostics.add(createFileDiagnostic(file, Debug.checkDefined(pos), Debug.checkDefined(end) - pos, diagnostic.diagnostic, ...diagnostic.args || emptyArray));
-            case FilePreprocessingDiagnosticsKind.ResolutionDiagnostics:
-                return diagnostic.diagnostics.forEach(d => programDiagnostics.add(d));
-            default:
-                Debug.assertNever(diagnostic);
-        }
-    });
-
     verifyCompilerOptions();
     performance.mark("afterProgram");
     performance.measure("Program", "beforeProgram", "afterProgram");
     tracing?.pop();
 
     return program;
 
+    function updateAndGetProgramDiagnostics() {
+        if (lazyProgramDiagnosticExplainingFile) {
+            // Add file processingDiagnostics
+            fileProcessingDiagnostics?.forEach(diagnostic => {

```

Comment:
I was going to ask if this should be set to undefined, but I assume not given it's passed to the next program?

I was just thinking of a case where we get the diags, then do more work, then end up duplicitively adding here, but I guess that can't happen as the assumption is that `lazyProgramDiagnosticExplainingFile` and `fileProcessingDiagnostics` are pushed to in one go?

Reply:
fileProcessingDiagnostics are generated only when constructing program and are needed when program is completely reused.